src/main/java/com/mattworzala/canary/server/givemeahome/StructureWriter.java

```diff
@@ -0,0 +1,91 @@
+
+public class StructureWriter {
+
+    public static void writeStructure(Structure structure, Path filePath) {
+        GsonBuilder gson = new GsonBuilder().excludeFieldsWithoutExposeAnnotation();
+
+        // when registering type adapters for specific generic values
+        // make sure that you register in the order of general -> specific
+
+        Type blockMapType = new TypeToken<Map<Integer, Block>>() {
+        }.getType();
+        gson.registerTypeAdapter(blockMapType, new BlockMapSerializer());
+
+        Type blockDefList = new TypeToken<List<Structure.BlockDef>>() {
+        }.getType();
+        gson.registerTypeAdapter(blockDefList, new BlockDefListSerializer());
+
+        gson.registerTypeAdapter(Vec.class, new VecSerializer());
+
+        var output = gson.create().toJson(structure);
+        File file = filePath.toFile();
+
+        try (FileOutputStream fos = new FileOutputStream(file);
+             BufferedOutputStream bos = new BufferedOutputStream(fos)) {
+            byte[] bytes = output.getBytes();
+            bos.write(bytes);
+        } catch (IOException e) {
+            e.printStackTrace();
+        }
+        System.out.println("JSON OUTPUT");
+        System.out.println(output);
+    }
+
+    private static class VecSerializer implements JsonSerializer<Vec> {
+        public JsonElement serialize(Vec vec, Type typeOfSRc, JsonSerializationContext context) {
+            return context.serialize(new int[]{vec.blockX(), vec.blockY(), vec.blockZ()});
+        }
+    }
+
+    private static class BlockDefListSerializer implements JsonSerializer<List<Structure.BlockDef>> {
+        public JsonElement serialize(List<Structure.BlockDef> blockDefList, Type typeOfSRc, JsonSerializationContext context) {
+
+            return new JsonPrimitive(blockDefList.stream()
+                    .map((blockDef) -> blockDef.blockId() + "," + blockDef.blockCount())
+                    .collect(Collectors.joining(";")));
+        }
+    }
+
+    private static class BlockMapSerializer implements JsonSerializer<Map<Integer, Block>> {
+        public JsonElement serialize(Map<Integer, Block> blockMap, Type typeOfSRc, JsonSerializationContext context) {
+            JsonArray arr = new JsonArray();
+            int index = 0;
+            while (blockMap.containsKey(index)) {
+                Block b = blockMap.get(index);
+                arr.add(new BlockSerializer().serialize(b, Block.class, context));
+                index++;
+            }
+            return arr;
+        }
+    }
+
+    private static class BlockSerializer implements JsonSerializer<Block> {
+        public JsonElement serialize(Block block, Type typeOfSrc, JsonSerializationContext context) {
+            JsonObject object = new JsonObject();
+            object.add("block", new JsonPrimitive(block.name()));
+            if (block.handler() != null) {
+                object.add("handler", new JsonPrimitive(block.handler().toString()));
```

Comment:
If I want to be real nitpicky, you should use the `context` object to handle creating the primitive objects. It handles a case where the serializer is modified, and in case gson ever makes breaking changes to `JsonPrimitive` for some reason.